dissect/target/plugins/os/windows/credhist.py

```diff
@@ -0,0 +1,211 @@
+    decrypted: bool = False
+
+    def decrypt(self, password_hash: bytes) -> None:
+        """Decrypt this CREDHIST entry using the provided password hash. Modifies ``CredHistEntry.sha1``
+        and ``CredHistEntry.nt`` values.
+
+        If the decrypted ``nt`` value is 32 characters we assume the decryption was successful.
+
+        Args:
+            password_hash: bytes of SHA1 password hash digest
+
+        Raises:
+            ValueError: If the decryption seems to have failed.
+        """
+        data = self.cipher_alg.decrypt_with_hmac(
+            data=self.raw.encrypted,
+            key=derive_password_hash(password_hash, self.user_sid),
+            iv=self.raw.pSalt,
+            hash_algorithm=self.hash_alg,
+            rounds=self.raw.dwPbkdf2IterationCount,
+        )
+
+        sha_size = self.raw.dwShaHashSize
+        nt_size = self.raw.dwNtHashSize
+
+        sha1 = data[:sha_size].hex()
+        nt = data[sha_size : sha_size + nt_size].rstrip(b"\x00").hex()
+
+        if len(nt) != 32:
+            raise ValueError("Decrypting failed, invalid password hash?")
+
+        self.decrypted = True
+        self.sha1 = sha1
+        self.nt = nt
+
+
+class CredHistFile:
+    def __init__(self, fh: BinaryIO) -> None:
+        self.fh = fh
+        self.entries = list(self._parse())
+
+    def __repr__(self) -> str:
+        return f"<CredHistFile fh='{self.fh}' entries={len(self.entries)}>"
+
+    def _parse(self) -> Iterator[CredHistEntry]:
+        self.fh.seek(0)
+        try:
+            while True:
+                entry = c_credhist.entry(self.fh)
+
+                # determine size of encrypted data and add to entry
+                cipher_alg = CipherAlgorithm.from_id(entry.algCrypt)
+                enc_size = entry.dwShaHashSize + entry.dwNtHashSize
+                enc_size += enc_size % cipher_alg.block_length
+                entry.encrypted = self.fh.read(enc_size)
+
+                yield CredHistEntry(
+                    version=entry.dwVersion,
+                    guid=UUID(bytes_le=entry.guidLink),
+                    user_sid=read_sid(entry.pSid),
+                    hash_alg=HashAlgorithm.from_id(entry.algHash),
+                    cipher_alg=cipher_alg,
+                    sha1=None,
+                    nt=None,
+                    raw=entry,
+                )
+        except EOFError:
+            pass
+
+    def decrypt(self, password_hash: bytes | str) -> None:
+        """Decrypt a CREDHIST chain using the provided password SHA1 hash."""
+        if isinstance(password_hash, str):
+            password_hash = bytes.fromhex(password_hash)
+
+        for entry in reversed(self.entries):
+            try:
+                entry.decrypt(password_hash)
+            except ValueError as e:
+                log.warning("Could not decrypt entry %s with password %s", entry.guid, password_hash.hex())
+                log.debug("", exc_info=e)
+                continue
```

Comment:
Took me a bit to really dig into what's happening here. You call `CredHistFile.decrypt` multiple times with different passwords. For some of those, encryption will work, for some it will fail. I assume you keep on trucking on an error in the case that you may have a password in your set that will successfully decrypted a _later_ password? (Does that even make sense? Presumably once you successfully decrypt the first in the chain, the rest should decrypt just fine?)

Wouldn't a flow like this make it a bit more clear what's happening (plus a comment about this behaviour in the docstring):
```
for entry in reversed(self.entries):
    if not entry.decrypted:
        entry.decrypt(password_hash)  # Let it raise the error - decryption of the file failed
    password_hash = entry.sha1
```

Reply:
That's a very good point, looks like I wasn't paying enough attention after all. I was very confused on the `password_hash = entry.sha1`  line below it, but for some reason I was completely blind to the `continue` in the `except` branch.

Apologies :wink: